python_modules/dagster/dagster/_core/execution/context/invocation.py

```diff
@@ -59,13 +58,63 @@
 
 def _property_msg(prop_name: str, method_name: str) -> str:
     return (
-        f"The {prop_name} {method_name} is not set on the context when a solid is directly invoked."
+        f"The {prop_name} {method_name} is not available on the context when an op or asset is"
+        " directly invoked."
     )
 
 
-class UnboundOpExecutionContext(OpExecutionContext):
-    """The ``context`` object available as the first argument to a solid's compute function when
-    being invoked directly. Can also be used as a context manager.
+class DirectExecutionContext:
+    """Helper class for direct invocation context classes."""
+
+    @property
+    def dagster_run(self) -> DagsterRun:
+        raise DagsterInvalidPropertyError(_property_msg("pipeline_run", "property"))
```

Comment:
Yep both PRs eliminate duplication, the difference is the method: factoring out (this PR) vs merging (my PR). Interested for @alangenfeld's take here. I went with merging because I didn't see independent use of bound/unbound execution context, so the simplicity of a single class made sense to me.

Reply:
I think the question is 
* do we care about mutating "bind time" information in place or should we chain objects? If we chain objects do we get value out of typing them differently? 

The initial design pass here was "immutable > mutable" & "ya different types align with mental model" 

If you look at how the events work, the list gets passed down and appended to so that some mutability we seem cool with and unclear we would trade away. 

```
context = build_op_context()
my_op(context)
context.op_def # does it matter if this works due to mutability?

other_op(context) # is re-using a context considered good?

Reply:
I think the more complex "bind time" changes at [3] maybe inform a preference on how to manage the context objects pre/post invocation. Maybe we go chained objects of a single type instead of mutate in place single type?

Reply:
>Maybe we go chained objects of a single type instead of mutate in place single type?

I like this. Reusing the same context object inside multiple invocations raises my hackles, I wasn't thinking of it but now I am seeing your example above.

If we do this, do you want to implement it here? Or maybe you just merge what you have minus the `DirectExecutionContext` factor, then I'll stack on you and modify my PR to have `bind` copy the object?

Reply:
@alangenfeld ping on ^^

Reply:
Ok I rebased this and took out the `DirectExecutionContext` changes quick. 

Its not clear to me what sequencing makes sense or if the changes can just be decoupled.

Reply:
>Ok I rebased this and took out the DirectExecutionContext changes quick.

Cool, I think then you should just merge this and I will merge after you, modifying my PR to have `bind` copy the object.

Reply:
Ok, I just bumped in to an issue with the assets binding https://github.com/dagster-io/dagster/pull/15622/files#r1281200178 so thats good motivation for me to see this through in some form